src/docfx/lib/JsonUtililty.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.Docs
+{
+    /// <summary>
+    /// Provide Utilities of Json
+    /// </summary>
+    public static class JsonUtililty
+    {
+        private static readonly ThreadLocal<JsonSerializer> s_defaultSerializer = new ThreadLocal<JsonSerializer>(
+            () => new JsonSerializer
+            {
+                NullValueHandling = NullValueHandling.Ignore,
+                ReferenceLoopHandling = ReferenceLoopHandling.Serialize,
+                Converters =
+                {
+                    new StringEnumConverter { CamelCaseText = true },
+                },
+                ContractResolver = new CamelCasePropertyNamesContractResolver(),
+            });
+
+        private static readonly JsonMergeSettings s_jsonMergeSettings = new JsonMergeSettings
+        {
+            MergeArrayHandling = MergeArrayHandling.Replace,
+        };
+
+        /// <summary>
+        /// Serialize an object to TextWriter
+        /// </summary>
+        public static void Serialize(TextWriter writer, object graph, bool isStable = false, Formatting formatting = Formatting.None, JsonSerializer serializer = null)
+        {
+            var localSerializer = serializer ?? s_defaultSerializer.Value;
+            if (isStable)
+            {
+                localSerializer.Formatting = Formatting.Indented;
```

Comment:
Do not mutate the state of localSerializer. you can use `JsonConvert.SerializeObject`

Reply:
good point, but `JsonConvert.SerializeObject` can not a take `TextWriter` but only `Object`.
will reset `localSerializer.Formatting` after using it

---
In reply to: [180296892](https://github.com/dotnet/docfx/pull/2609#discussion_r180296892) [](ancestors = 180296892)